### ISSUE TITLE
RequestCacheConfigurer default request matcher doesn't allow saving POST requests when csrf is enabled

### ISSUE THREAD

**Issue by codependent** [spring-projects/spring-security#7365]
### Summary
In a web application with Spring Security 4.2.x RequestCacheConfigurer configures a default request matcher that doesn't allow saving POST requests when csrf is enabled.

### Actual Behavior

When we submit a POST request against a secured endpoint the application redirects to the login page. After successful login the original POST request isn't restored so the invocation fails.

### Expected Behavior

The original POST request is restored.

### Configuration

I'm relying on Spring Security default configurations for the request cache. The key part is here `RequestCacheConfigurer` : 

```
	private RequestMatcher createDefaultSavedRequestMatcher(H http) {
                ...
                if (isCsrfEnabled) {
			RequestMatcher getRequests = new AntPathRequestMatcher("/**", "GET");
			matchers.add(0, getRequests);
		}
```

If csrf is enabled only get requests will be cached.

Is there any security related reason why this matcher isn't supporting POST invocations?

### Version

4.2.12

**Comment by jzheaux** [comment]
I believe it simply wouldn't work. Since the CSRF token is bound to the session and the session is invalidated at authentication time, then the token stored as part of the post would fail to match.

**Comment by codependent** [comment]
Yes, but that POST url could use a csrf exclusion. In that case if (isCsrfEnabled) is still true but the POST request can’t be saved.

**Comment by jzheaux** [comment]
Yes, that's a good point. The `RequestCacheConfigurer` could possibly consult `CsrfConfigurer`'s `RequestMatcher` to improve its own decision making. I'll take a look at this.

In the meantime, would this work for you:

```java
@Bean 
RequestCache requestCache() {
    HttpSessionRequestCache cache = new HttpSessionRequestCache();
    cache.setRequestMatcher(myRequestMatcher());
    return cache;
}
```

where `myRequestMatcher()` is a request matcher that matches what your `csrf()` configuration will ignore?

**Comment by codependent** [comment]
Thank's @jzheaux, I did exactly what you suggest to work around it, but I'd be great if the framework provided that behaviour out of the box ;-)

**Comment by uuufo** [comment]
Hello! I'm looking to implement a solution like above while retaining the default matcher created by RequestCacheConfigurer in Spring Security 6.1.x:

```java
HttpSessionRequestCache cache = new HttpSessionRequestCache();
cache.setRequestMatcher(new OrRequestMatcher(
        new AntPathRequestMatcher("/post-endpoint", "POST"),
        // matcher created by RequestCacheConfigurer.createDefaultSavedRequestMatcher()
));
```

Currently I'm replicating the behavior of `createDefaultSavedRequestMatcher()`, but was wondering if there is a better way?

I've also noticed that this works without excluding the endpoint from CSRF. It seems after login the session ID is updated and the CSRF token is replaced, but then the saved request is first hit using GET (with ?continue) followed by the actual POST request. Does this cause it to be submitted with the new CSRF token?
